### ISSUE TITLE
Inform the user about the current project type somewhere on the UI

### ISSUE THREAD

**Issue by goanpeca** [spyder-ide/spyder#13286]
With the work in https://github.com/spyder-ide/spyder/pull/13280, now developers can register custom project types to enhance the capabilities of Spyder.

@nerohmot thinks it would be a good idea to display the current type of the project somewhere in the UI.

Some options include:

- Title bar (Red)
- Project Pane toolbar (Green)
- Status bar (Blue)

![image](https://user-images.githubusercontent.com/3627835/87430773-5129e100-c5ab-11ea-9710-451e9614e93a.png)

**Comment by jitseniesen** [comment]
It is not clear to me how important the project type is. What can the new project types do? I assume that the project type determines which files the project is populated with originally, but are there any other effects?

**Comment by goanpeca** [comment]
>It is not clear to me how important the project type is. What can the new project types do? I assume that the project type determines which files the project is populated with originally, but are there any other effects?

A project type for now defines `open` `close` and `create` actions, open and close could add new items on different places of the UI.

**Comment by ccordoba12** [comment]
> A project type for now defines open close and create actions, open and close could add/hide items on different places of the UI.

Could you expand on this by giving a concrete example?

**Comment by nerohmot** [comment]
Until now we have only one project type, I call it "python un-managed".
  - Python = the language (now that LSP is extended others can follow)
  - un-managed = one populates the project as one likes

Together with Gonzalo I am implementing the "ATE plugin" that provides the "ATE Project" type, it is what I would call "python ATE managed"
  - Python = the language
  - ATE managed = one populates/maintains the project by a series of wizards (instead of using the "new module/package/file/folder context menus in the project pane)

It stands to reason that now that we have more than one project type, it should be marked somewhere with what kind of project type we are working ... (initial question from Gonzalo)

Title bar (Red), already displaying the project name, and I think that suffices.
Project Pane toolbar (Orange) that is a good place to indicate the project type (this space is empty now anyway) as Gonzalo notes, the project pane can be hidden (and in ATE Project we will do so, there is auto generated files that we don't want the user to mess with) never the less, if one enables the project pane it is good to see the type!
Status bar (Blue) that is a perfect place that is always visible.

Conclusion: I think the project type should be visible in both the project pane "title" **AND** the status bar.

**Comment by jitseniesen** [comment]
Thanks for your explanation. As I understand it, having an ATE Project open will be a considerable change in how Spyder is used, in that users should not add files willy-nilly to the project but instead use the wizards provided by the ATE plugin. Therefore you want to have an indicator of the project type. I guess the ATE pane itself indicates that an ATE project is open, but it may be hidden. So I get why you want an indicator in the status bar.

On the other hand, many users will not have the ATE plugin installed and they will thus have (at least at present) only one project type, which means that it is not necessary to have an indicator. As the screen is already cluttered enough as it is, in this case I would prefer not to have any indicator of project type in the status bar.

I can think of the following possibilities which achieves both goals:
1. Spyder puts the project type in the status bar. If only one project type is defined, then Spyder does not put anything in the status bar.
2. Spyder puts the project type in the status bar. If the default project type is selected, then Spyder does not put anything in the status bar. This is only a small change from 1.
3. The ATE plugin is responsible for putting the project type in the status bar. In this case, the plugin writers can decide themselves what is best for the situation that they have in mind. I don't know how hard this is technically, but I think it would be a good option in general to allow plugins to add indicators to the status bar.

I am less bothered about what happens to the project pane title. It's acceptable for me to always display the project type there, but it also makes sense to do the same as for the status bar in order to be consistent.

Incidentally, I am not sure that we need different project type for different programming languages. This would make it more difficult to mix different languages. But this is not relevant for the current issue.

**Comment by CAM-Gerlach** [comment]
I agree with @jitseniesen . If and when custom project types are more common, we can reconsider, but given we estimate Spyder is used by somewhere on the order of 10^5 and 10^6 worldwide, and the currently sole proposed plugin, still under development, that implements a non-default project type will likely be in use by 10^1-10^2 users, adding extra visual clutter to the remaining 99.9-99.999% of Spyder users does not appear justified. However, Jitse's suggestions above (particularly option 2) appear to solve the problem without incurring a significant tradeoff for the use cases for which a project type becomes meaningful.

> I don't know how hard this is technically, but I think it would be a good option in general to allow plugins to add indicators to the status bar.

If I recall our discussions accurately, this should be possible with the new plugin architecture in Spyder 5 that @goanpeca is developing as part of this project. Is that right, @goanpeca ?

> I am less bothered about what happens to the project pane title. It's acceptable for me to always display the project type there, but it also makes sense to do the same as for the status bar in order to be consistent.

The project pane title is certainly much less prime real estate, but I would favor being consistent with the status bar and avoiding extra clutter if the project type is the default (which also avoids us having to come up with a fancy name for it :smile: )

**Comment by goanpeca** [comment]
>On the other hand, many users will not have the ATE plugin installed and they will thus have (at least at present) only one project type, which means that it is not necessary to have an indicator. As the screen is already cluttered enough as it is, in this case I would prefer not to have any indicator of project type in the status bar.

Spyder5 will already include a SpyderProject Plugin (based on cookie-cutter) so let's assume for this discussion that there will be more than one project type for sure.

**Comment by CAM-Gerlach** [comment]
Thanks for the quickly reply, @goanpeca !

> @jitseniesen and @CAM-Gerlach Spyder5 will already include a SpyderProject Plugin Type (based on cookie-cutter) so let's assume for this discussion that there will be more than one project type for sure.

True, though the percentage of users using Spyder to develop a Spyder plugin is also likely to be a very small fraction of the total userbase. As more project types are added that will hopefully change, but a large fraction if not the majority of users are likely to be still using the default project, as many if not most Spyder users today don't even use projects at all—and if no project is open, we'd naturally not want to show the statusbar item either. As such, I'm thinking Jitse's option 2 might make the most sense, since if users are actually using a project type Spyder will show it, while if they are using the default type (which has no special effect) it will not. If and when a larger minority of users are using a project type, then maybe it would make more sense to show it even if it is the default...I'm not sure.

> A project type is currently very simple as it provides a create/open/close methods. 

I understand the create methods, but could you explain what the custom open/close methods do? Do you mean file open/close or project open/close? In @nerohmot 's example, what would that correspond to?

**Comment by goanpeca** [comment]
open/close are actions to perform when a project that has been created already is loaded. Close is similar. 

In the example, opening a project will do a lot of stuff under the hood to display things in the extra Panel + show or hide some toolbars or panes. Closing would probably just hide the panel and extra toolbar.

**Comment by CAM-Gerlach** [comment]
Okay, thanks. Its clear now; my confusion was over the unclear object of the actions (a file vs. the project). Similarly, I'm guessing the the `create` action refers to creation of the whole project (e.g. setting up a custom cookiecutter), as opposed to creating a new files within a project (via a custom wizard/individual file cookiecutter)?

**Comment by goanpeca** [comment]
>Okay, thanks. Its clear now; my confusion was over the unclear object of the actions (a file vs. the project). Similarly, I'm guessing the the create action refers to creation of the whole project (e.g. setting up a custom cookiecutter), as opposed to creating a new files within a project (via a custom wizard/individual file cookiecutter)?

Correct.

A new project type could be a DjangoProject using a Cookiecutter under the hood to use the well known templates out there + maybe specifying (eventually) HTML/CSS/JS/Python Language servers.

**Comment by ccordoba12** [comment]
I agree with @jitseniesen that the best place to put the project should be the status bar.

**Comment by CAM-Gerlach** [comment]
@ccordoba12 As I understand it, @jitseniesen 's main point was that the project type _shouldn't_ be shown on the status bar unless it is actually distinct from the default, which doesn't do anything special. As far as I've gathered, it seems we all mostly agree agree that if there is a project with a specialized project type open, the status bar is probably the best all-around place to show it, but what is being discussed is whether to show anything if one is not (either no project is open, or it is of the default type that doesn't currently do anything).

So, @ccordoba12 What do you think of @jitseniesen 's proposal? @goanpeca and @nerohmot , would that be a good compromise or does it pose serious problems for your use case? If the latter, could an extra status-bar warning the plugin could inject on the statusbar that an ATS project is not activated be an equal or better solution? Thanks.

**Comment by goanpeca** [comment]
I do not consider that there will be a "default project". We had an empty project as a simple thing but it is pretty much useless and I foresee that Spyder 5 will already include more than just an empty or a Spyder Plugin one, so I do not see any value if giving the very lacking "Empty Project" any special treatment. I see all projects equal and I would not special case showing one or not. The question is more: Where to show it from the options (if any of those apply) or any other suggestion of location?

I think the status bar is too crowded but that is another discussion where items in there should be closer together and maybe a tad smaller.

**Comment by jitseniesen** [comment]
The problem is, I have no idea what a project type is supposed to do, except for what is explained for the ATE managed project type. If a project type only sets up some files when creating the project and does not do anything else (which is what I imagine you mean with projects based on cookiecutter), then I see no need to display the project type.

**Comment by CAM-Gerlach** [comment]
Unfortunately, in my experience, most of the Spyder users I know don't use projects at all, and those IRL and on here that do, often report they just use them for their workspace/session-switching features or work on multiple related Python projects in one Spyder project (including myself at times), and many if not most scientific Python users (unlike us) unfortunately mostly work interactively or on scripts rather than on Python packages or similar structures that these project types seem designed for. As such, I'd be surprised if users who A. Use projects, B. Use them for more than just session/workspace management and C. Use a special project type, will ever constitute a majority or even more than modest minority of Spyder users.

@goanpeca I understand you don't want to special-case an empty project, but since it is more the general/null case that doesn't involve any "special treatment" of creating, opening and closing projects, and the only one a large majority of users are likely to make much use of, I think it deserves consideration here.

Also, just to be clear, what do you propose to do if no project is open? You do agree with not showing the project type statusbar item in that scenario, right?

> If a project type only sets up some files when creating the project and does not do anything else

I suppose different project types could do things like show the appropriate window layouts and panes for a Python package, Django web app, Spyder plugin, interactive use, etc. (basically automatically switching the existing window layout presets) or launching a console with certain default settings (e.g. a Cython project type could load Cython by default in the IPython console). However, on the other hand I'm not sure how big of a value add these are for most use cases when these could be done in a few clicks with existing features, as opposed to relatively radical overhauls like what is being done for the ATE project type.

**Comment by goanpeca** [comment]
> Also, just to be clear, what do you propose to do if no project is open? You do agree with not showing the project type statusbar item in that scenario, right?

If no project is open why would we show a project type 🙃 ?

>as opposed to relatively radical overhauls like what is being done for the ATE project type.

Not really radical.

**Comment by nerohmot** [comment]
Indeed : no project open, nothing to show, project open show the type :-)

**Comment by jitseniesen** [comment]
> I do not see any value if giving the very lacking "Empty Project" any special treatment. I see all projects equal and I would not special case showing one or not. The question is more: Where to show it from the options (if any of those apply) or any other suggestion of location?

I see the blank canvas of "Empty Project" as a feature :smile: In fact, I think it is exactly what I need for most of my use cases.

If we don't want to special case "Empty Project", then how about adding a field to the project type definition which specifies whether the project type should be displayed in the status bar? Based on what is said here, I don't see why the SpyderPlugin project type needs to be displayed prominently. It can be displayed somewhere in the Project pane, or in a project info dialog or preference page which I assume will be added at some point.
